pets-app/src/main/java/com/example/pets/odata/datasource/PetsAppDataSource.java

```
@@ -0,0 +1,132 @@
+import static com.example.pets.model.odata.util.EdmConstants.ENTITY_SET_NAME_DOGS;
+import static com.example.pets.model.odata.util.EdmConstants.ENTITY_SET_NAME_USERS;
+import static com.example.pets.odata.context.PetsAppContext.getApplicationContext;
+import static com.example.pets.odata.util.ODataConstants.ID;
+
+public class PetsAppDataSource implements DataSource {
+
+    private final ODataServiceProvider oDataServiceProvider;
+
+    public PetsAppDataSource() {
+        ODataCatService catService = getApplicationContext().getBean(ODataCatService.class);
+        ODataUserService userService = getApplicationContext().getBean(ODataUserService.class);
+        ODataDogService dogService = getApplicationContext().getBean(ODataDogService.class);
+        ODataAnimalService animalService = getApplicationContext().getBean(ODataAnimalService.class);
+        this.oDataServiceProvider =
+                new ODataServiceProvider(catService, dogService, animalService, userService);
+    }
+
+    @Override
+    public List<?> readData(EdmEntitySet edmEntitySet) throws ODataNotFoundException, EdmException {
+        String entitySetName = edmEntitySet.getName();
+        return oDataServiceProvider
+                .getODataService(entitySetName)
+                .findAll();
+    }
+
+    @Override
+    public Object readData(EdmEntitySet edmEntitySet, Map<String, Object> map) throws ODataNotFoundException, EdmException {
+        String entitySetName = edmEntitySet.getName();
+        Long firstLayerEntityId = (Long) map.get(ID);
+        return oDataServiceProvider
+                .getODataService(entitySetName)
+                .findById(firstLayerEntityId);
+    }
+
+    @Override
+    public Object readData(EdmFunctionImport edmFunctionImport, Map<String, Object> map, Map<String, Object> map1) throws ODataNotImplementedException, ODataNotFoundException, EdmException, ODataApplicationException {
+        return null;
```

Don't return `null`. Never. Better to throw an exception